### ISSUE TITLE
How to handle the documentation access from readium.github.io

### ISSUE THREAD

**Issue by camill-a** [readium/readium-2#34]
The 2 solutions we came up with, in chronological order, are:

# include_remote
The first solution is a custom Jekyll plugin called `remote-include`.
It injects ressource which are outside of the local Jekyll repository (i.e. the repo where Jekyll is configured; vanilla `include`/`include_relative` can only access local ressources).

Pros :
- Simple and easy to change, if something less hackish comes up in the future.
- Documentation repo & Jekyll repo are uncoupled.

Cons :
- No site regeneration on remote files modification, hence the requirement to "touch" any file in the Jekyll repo to force a regeenration of the Jekyll site.
- Each time we add a new documentation file, we would have to add a new page/post like:
```
---
layout: post-using-remote-include
title:  "Readium2 Streamer Roadmap"
date:   2017-03-23 11:36:00 +0100
categories: overview
permalink: /r2-streamer-roadmap/
---
{% remote_include https://raw.githubusercontent.com/readium/readium-2/master/streamer/roadmap.md %}
```
The code can be checked [here](https://github.com/readium/readium.github.io/tree/develop) on the develop branch of the readium site repo. You can also build it locally to see it live using [bundler](https://help.github.com/articles/setting-up-your-github-pages-site-locally-with-jekyll/).

# gh-pages
This is the second solution we came up with, trying to tackle the "auto regeneration" problem cited above.
_This solution hasn't been fully implemented yet since it will change the architecture of the repo and the content of the files._
We create a `gh-pages` branch on the `readium-2` documentation repository.
Inside it we add the 'Jekyll conf+ressources' (/_includes/*, /_layouts/*,_config.yml, /assets/*).
We add a 'front matter' to each documentation file in order for Jekyll to process them (read convert `.md` to `.html`).

Pros :
- When you change a documentation file, the website is regenerated (refreshed) automatically.
- When you create a new documentation file it directly appears on the website.

Cons :
- Jekyll config files must be duplicated and synced (or a common jekyl conf repo must be created as in the Spring project [see here](https://github.com/spring-projects/gh-pages)).
- The documentation is handled in a  'gh-pages' branch, not the master branch (which will be deleted).
- All doc pages must get a front matter (i.e. jekyll header).
 
[Here](https://github.com/readium/readium-2/tree/gh-pages) is the branch we made tests with.

Which solution would you like to see implemented?

**Comment by llemeurfr** [comment]
IMHO the second solution is much better, since it's much more dynamic and the work required (adding front matters to the docs) is minimal. 
The only burden is the copy of the jekyll config in the readium-2 space: low burden since it won't change in the medium term.

**Comment by llemeurfr** [comment]
We don't even have to move to a gh-pages branch, we just need to update the repo settings to use the master branch for github pages.

**Comment by HadrienGardeur** [comment]
What I'm really wondering about is mostly timing. 
We're still in the early days of handling all of this documentation and we're likely to iterate over this repo for some time.

IMO rolling out an HTML version only makes sense when we're a little bit more stable than that.

**Comment by HadrienGardeur** [comment]
Also, there's one thing that I haven't seen mentioned here: should the Readium-2 docs remain separate from the rest of the Readium docs already generated using Jekyll?

We could only publish documents that we're confident about in [readium/readium.github.io](https://github.com/readium/readium.github.io) using `remote_include` and this mechanism could also be useful for maintaining the LCP spec.

**Comment by camill-a** [comment]
It's possible using the solution 2, by no putting front-matters in the documents we don't want to publicly expose.

**Comment by HadrienGardeur** [comment]
That's true, but in that case I'd like to see consistency across all repos dedicated to documentation (LCP and Web Publication Manifest too).

As for the pros and cons, I have mixed feelings. 
For a specification document (Web Publication Manifest or LCP) having to manually touch a document to trigger a regeneration is a feature, not a bug IMO. I see the additional step as a safety net to avoid versions that we don't want to get published.

**Comment by llemeurfr** [comment]
In the Github spirit, I would therefore treat the LCP spec using a develop branch (moving) and a master branch which contains the tagged revisions. Only the master branch would be seeable as html doc (thanks to its jekyll front matter).

**Comment by HadrienGardeur** [comment]
I'm not a big fan for documents because it means lower visibility, which is the opposite of what we want (we need as many eyes as possible to review a draft).

I like the W3C approach better where you have both the current draft and "publishing snapshots" in the same branch: https://github.com/w3c/dpub-pwp

**Comment by danielweck** [comment]
Can Jekyll "rendering" be manually triggered for a release tag on the master branch (for official stable documentation updates), as well as automatically triggered (continuous integration) for editors drafts?
What URL / domain would be used for stable docs vs. living documents?
e.g.
https://readium.org/lcp/specs/1.0
vs.
https://readium.org/lcp/specs/draft

**Comment by llemeurfr** [comment]
@danielweck I don't think so. Jekyll is triggered by a modification on a file in the selected branch (gh-pages or master). It's a basic mechanism.

**Comment by danielweck** [comment]
apparently the Jekyll build scripts can be invoked automatically from a continuous integration server following a commit+push on any branch (which can be filtered via a whitelist based on a syntactical convention, or just a complete branch name):
https://jekyllrb.com/docs/continuous-integration/travis-ci/

**Comment by llemeurfr** [comment]
Now that Ric intends to move readium.org and epubzone to githib/jekyll, it would be good to be able to close this issue. 

The gh-pages solution works, but imposes a copy of the jekyll readium templates in the r2 repo, which may make maintenance more difficult for these templates. 

The best conclusion therefore seems to be using the first solution (the ad-hoc "include-remote" script) for R2 including selected architecture texts in readium.org. 

I'll close this issue after validation in a R2 call.

**Comment by llemeurfr** [comment]
In fact in https://help.github.com/articles/adding-jekyll-plugins-to-a-github-pages-site/ I discover that plugin like the one we tried to import external md pages (from the readium-2 repo) to the readium.org webpages are not allowed by github (for security reasons), i.e. they work locally (when using bundle exec jekyll serve) but not when github itself is the engine (the solution is to copy to github the "resolved" site, which is not what we want). We were able to commit on the develop branch with no issue, but it gives an error when merged in the master branch. 

We currently want to pick and choose some pages from the readium-2 repo and have them accessible on the readium.org website: the best solution I see for the short term is a copy of the .md content. 
I'll do it now, so the readium.github.io website can be up to date.
